lib/transport/selenium-webdriver/method-mappings.js

```diff
@@ -934,14 +935,171 @@ module.exports = class MethodMappings {
+          const cdpConnection = await cdp.getConnection(this.driver);
+
+          await cdpConnection.execute(
+            'Emulation.setDeviceMetricsOverride',
+            metrics
+          );
+
+          return {
+            value: null
+          };
+        },
+
+        async interceptNetworkCalls(userCallback) {
+          const cdpConnection = await cdp.getConnection(this.driver);
+
+          cdpConnection._wsConnection.on('message', (message) => {
+            const params = JSON.parse(message);
+            if (params.method === 'Network.requestWillBeSent') {
+              const requestParams = params['params'];
+
+              userCallback(requestParams);
+            }
+          });
+
+          await cdpConnection.execute(
+            'Network.enable',
+            {}
+          );
+
+          return {
+            value: null
+          };
+        },
+
+        async mockNetworkResponse(urlToIntercept, response) {
+          const cdpConnection = await cdp.getConnection(this.driver);
+
+          const {status=200, headers: headersObject, body: bodyPlain=''} = response;
+
+          const headers = [];
+          if (headersObject) {
+            for (const [name, value] of Object.entries(headersObject)) {
+              headers.push({name, value});
+            }
+          }
+
+          // Convert body to base64
+          const bodyBase64 = Buffer.from(bodyPlain, 'utf-8').toString('base64');
+          
+          cdpConnection._wsConnection.on('message', (message) => {
+            const params = JSON.parse(message);
+            if (params.method === 'Fetch.requestPaused') {
+              const requestPausedParams = params['params'];
+              if (requestPausedParams.request.url === urlToIntercept) {
+                cdpConnection.execute('Fetch.fulfillRequest', {
+                  requestId: requestPausedParams['requestId'],
+                  responseCode: status,
+                  responseHeaders: headers,
+                  body: bodyBase64
+                });
+              } else {
+                cdpConnection.execute('Fetch.continueRequest', {
+                  requestId: requestPausedParams['requestId']
+                });
+              }
+            }
+          });
+      
+          await cdpConnection.execute(
+            'Fetch.enable',
+            {}
+          );
+          await cdpConnection.execute(
+            'Network.setCacheDisabled',
+            {cacheDisabled: true}
+          );
+
+          return {
+            value: null
+          };
+        },
+
+        async enablePerformanceMetrics(enable) {
+          // Disable the metrics once even before enabling it.
+          await this.driver.sendAndGetDevToolsCommand('Performance.disable');
+
+          if (enable) {
+            await this.driver.sendAndGetDevToolsCommand('Performance.enable');
+          }
+
+          return {
+            value: null
+          };
+        },
```

Comment:
If we are disabling and enabling performance metrics before every command. Shouldn't this be hidden like every time we call any command-related performance we should enable it. Is there any added benefit of providing this as a client command?

Reply:
`Performance.getMetrics` CDP command returns the performance metrics collected starting from the time `Performance.enable` was called. So we need to enable the performance metrics collection before navigating to a website and ask for the performance metrics after navigating to the website.

This is the same reason I'm disabling it before enabling, so that if the user want to start afresh with the metrics collection, they can just call `browser.enablePerformanceMetrics` instead of first calling something like `browser.disablePerformanceMetrics` and then `browser.enablePerformanceMetrics` in the very next line.

Although the working of these commands is not documented anywhere, so all the above inferences are strictly based on experiments done by me.